### ISSUE TITLE
Release 4.0.0

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#1117]
Can be labelled **ready** when all other issues on this milestone are closed.

- [x] check integration package compatibility
  - [x] [AutoFixture.AutoFakeItEasy](https://github.com/AutoFixture/AutoFixture) - @blairconrad 
  - [x] [AutoFixture.AutoFakeItEasy2](https://github.com/AutoFixture/AutoFixture) - @blairconrad
  - [x] [Machine.Fakes.FakeItEasy](https://github.com/machine/machine.fakes) - @blairconrad
    **Broken**. See https://github.com/FakeItEasy/FakeItEasy/issues/1117#issuecomment-317392975. Created https://github.com/machine/machine.fakes/issues/40.
  - [x] [Autofac.Extras.FakeItEasy](https://github.com/autofac/Autofac.Extras.FakeItEasy) - @thomaslevesque
    Works, but package downgrade warnings during restore. Should work with no issues if upgraded to NetStandard.Library 1.6.1. Consuming code that uses both Autofac.Extras.FakeItEasy and FakeItEasy 4.0.0-beta001 works fine.
  - [x] [EntityFrameworkTesting.FakeItEasy](https://github.com/scott-xu/EntityFramework.Testing) - @blairconrad
    - There's been no update  to the project since we released 3.0.0, and we haven't changed anything to improve upon what we reported then:
        - Fails. `IReturnValueConfiguration.ReturnsLazily` signature changed. On the other hand, the package is restricted to >= 1.25.2 && < 2.0.0, so I don't think we need to worry about it. 
        - Related issue: https://github.com/scott-xu/EntityFramework.Testing/issues/17
- [x] ~~run code analysis in VS in *Release* mode and address all coding and style violations (send a regular PR which must be merged before continuing)~~
- [x] ensure completeness of draft [release notes](https://github.com/FakeItEasy/FakeItEasy/releases), including non-owner contributors, if any (move release notes forward from any pre-releases to the current release)
- change "vNext" names to appropriate x.y.z version for the next:
    - [x] GitHub Release
    - [x] milestone
    - [x] release checklist (this issue)
- create new GitHub artifacts for the next release:
    - [x] create a new draft GitHub Release with name "vNext"
    - [x] create a new milestone for the next release with name "vNext"
    - [x] create a new issue (like this one) with name "Release vNext" for the next release, adding it to the new milestone
- [x] create release-x.y.z branch on upstream
- [x] perform any last-minute tasks on release-x.y.z
- [x] submit a PR to merge the release-x.y.z branch to master (#1172)
- [x] ~~submit a PR to merge the release-x.y.z branch to develop, if there are any changes on release-x.y.z~~
- [x] once release-x.y.z has been merged, tag master and push to upstream. This will push the nupkgs to [nuget.org](https://www.nuget.org/profiles/FakeItEasy)
- [x] de-list pre-release or superseded buggy NuGet packages if present
- [x] attach all nupkgs from the tagged build to the GitHub Release
- [x] publish the GitHub release
- [x] tweet, mentioning contributors, and post link as comment here for easy retweeting ;-)
- [x] post tweet in [Gitter](https://gitter.im/FakeItEasy/FakeItEasy)
- [x] post a link to the GitHub Release in each issue in this milestone, with thanks to contributors
- [x] close this issue
- [x] close the [milestone](https://github.com/FakeItEasy/FakeItEasy/milestones?state=open) for this release

**Comment by blairconrad** [comment]
#1156 is almost fixed, and I don't see anything outstanding that feels "breaking". Should we try a 4.0.0 beta? RC? Beta first, then test the 3–4 packages that we like to make sure we didn't break, then RC?

**Comment by thomaslevesque** [comment]
Let's go for a beta 👍

**Comment by blairconrad** [comment]
It's been nearly two weeks and the world hasn't ended.  Shall we start thinking about releasing 4.0?
If we don't plan on putting anything else in, we could start testing the package with our favourite downstream packages.

**Comment by thomaslevesque** [comment]
The [only feedback I saw](https://www.thomaslevesque.com/2016/01/17/automatically-inject-fakes-in-test-fixture-with-fakeiteasy/#comment-89384) was about the deprecation of fixture initialization, and that's not something we're going to rollback anyway.

So yes, I think we're ready to plan the release. I can't think of anything else we should add.

**Comment by blairconrad** [comment]
Okay. Tentatively package to the checklist. Will start with same as last time. Of course, do amend if you think warranted.

**Comment by thomaslevesque** [comment]
I checked Autofac.Extras.FakeItEasy; it compiles without errors or warnings, and the tests pass, but there are many "package downgrade" warnings during restore, due to the fact that it references `NetStandard.Library` 1.6.0 whereas FakeItEasy references 1.6.1.

I also tried to check Machine.Fakes.FakeItEasy, but I haven't been able to build it. It uses the preview .NET Core SDK, so it requires VS2015, which I don't have anymore. The build scripts also don't work, I haven't investigated why.

**Comment by blairconrad** [comment]
Thanks, @thomaslevesque.

> I checked Autofac.Extras.FakeItEasy; it compiles without errors or warnings, and the tests pass, but there are many "package downgrade" warnings during restore, due to the fact that it references `NetStandard.Library` 1.6.0 whereas FakeItEasy references 1.6.1.

But this downgrade issue should be a problem only for consumers of Autofac.Extras.FakeItEasy, no? Do we worry about Autofac.Extras.FakeItEasy having to change their reference if they pick up FIE 4.0.0? I thought the concern was whether some project using Autofac.Extras.FakeItEasy who, when the upgraded their FakeItEasy, would encounter an error from Autofac.Extras.FakeItEasy. I feel bad asking this, but did you test by recompiling Autofac.Extras.FakeItEasy with the new FakeItEasy again? The main test is to swap the new FakeItEasy into a project using the current Autofac.Extras.FakeItEasy.

> I also tried to check Machine.Fakes.FakeItEasy, but I haven't been able to build it. It uses the preview .NET Core SDK, so it requires VS2015, which I don't have anymore. The build scripts also don't work, I haven't investigated why.

I can take a look.

**Comment by thomaslevesque** [comment]
> But this downgrade issue should be a problem only for consumers of Autofac.Extras.FakeItEasy, no?

Yes

> Do we worry about Autofac.Extras.FakeItEasy having to change their reference if they pick up FIE 4.0.0?

I don't think so

> I thought the concern was whether some project using Autofac.Extras.FakeItEasy who, when the upgraded their FakeItEasy, would encounter an error from Autofac.Extras.FakeItEasy. I feel bad asking this, but did you test by recompiling Autofac.Extras.FakeItEasy with the new FakeItEasy again? The main test is to swap the new FakeItEasy into a project using the current Autofac.Extras.FakeItEasy.

You're right, I always forget. I did recompile Autofac.Extras.FakeItEasy against FIE 4.0.0. I'll try what you said.

**Comment by thomaslevesque** [comment]
OK, it works as expected with the current version of Autofac.Extras.FakeItEasy and FIE 4.0.0-beta001. It even removes a warning:

> E:\tmp\TestAutofacFakeItEasy\TestAutofacFakeItEasy.csproj : warning NU1603: Castle.Core 4.0.0 depends on System.ComponentModel.TypeConverter (>= 4.0.1) but System.ComponentModel.TypeConverter 4.0.1 was not found. An approximate best match of System.ComponentModel.TypeConverter 4.1.0 was resolved.

**Comment by blairconrad** [comment]
Machines.Fakes.FakeItEasy has some incompatibilities. Four unit tests fail, with 2 different errors:

```
System.ArgumentException: GenericArguments[0], 'System.String', on 'System.Nullable`1[T] IsNull[T](FakeItEasy.IArgumentConstraintManager`1[System.Nullable`1[T]])' violates the constraint of type 'T'. ---> System.Security.VerificationException: Method FakeItEasy.ArgumentConstraintManagerExtensions.IsNull: type argument 'System.String' violates the constraint of type parameter 'T'.
```

occurs when Machine.Fakes tries to translate `Param<String>.IsNull` into `A<T>.That.IsNull` and finds the overload for a nullable struct instead of a class. (The former was added to 3.3.0(!) in issue #1062.)

It's Disappointing that we actually broke the package in 3.3.0 by adding an extension method!

Also:
```
System.InvalidOperationException: Unable to find target property Not on instance of target type FakeItEasy.IArgumentConstraintManager`1[[System.String, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]
```

similarly occurs when Machine.Fakes tries to translate a `Param<String>.IsNotNull`. We are moving `Not` with the  4.0.0 release, in #1063. (And we added `IsNotNull`, which would save Machine.Fakes a step, in 3.3.0 - #1062 again).

I don' think we should change our package based on these findings, but later today I can create an issue for Machine.Fakes so they can react to the new package.

**Comment by blairconrad** [comment]
https://twitter.com/fakeiteasyfx/status/890550729353699328
https://twitter.com/fakeiteasyfx/status/890551033910394880

**Comment by blairconrad** [comment]
holding on de-listing

**Comment by thomaslevesque** [comment]
Thanks for the release @blairconrad!

> holding on de-listing

Why?

**Comment by blairconrad** [comment]
I should've said. NuGet is real slow with the indexing today.

**Comment by blairconrad** [comment]
nuget.org finished indexing.
